controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -535,6 +617,28 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
 	return bootstrapRef, nil
 }
 
+func (r *KubeadmControlPlaneReconciler) failureDomainForUpgrade(cluster *clusterv1.Cluster, machines []*clusterv1.Machine) *string {
```

Comment:
I wish this and failureDomainForScaleDown and failureDomainForScaleUp had the same signature. Even if the contents are the same for now (they may be different), it would be nice if the ergonomics were the same and the caller was always required to pass in the list of machines.

Can be out of scope for this PR, no big deal.